File: Programs/Finding_3-Digit_Even_Numbers.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> findEvenNumbers(vector<int>& digits) {
        unordered_map<int, int> freq;
        for(const int& digit: digits) ++freq[digit];

        vector<int> ans;

        for(int num= 100; num < 1000; ++num) {
            int hundreds= num / 100;
            int tens= (num / 10) % 10;
            int units= num % 10;

            if(units % 2 != 0) continue;

            --freq[hundreds];
            if(freq[hundreds] >= 0) {
                --freq[tens];

                if(freq[tens] >= 0) {
                    --freq[units];

                    if(freq[units] >= 0) ans.push_back(num);

                    ++freq[units];
                }                
                ++freq[tens];
            }
            ++freq[hundreds];
        }

        return ans;
    }
};

File: Programs/Finding_3-Digit_Even_Numbers.py
class Solution:
    def findEvenNumbers(self, digits: List[int]) -> List[int]:
        freq= {i:0 for i in range(10)}
        for digit in digits:
            if digit not in freq.keys():
                freq[digit]= 1
            else:
                freq[digit] += 1
        
        ans= list()

        for num in range(100, 1000, 1):
            hundreds= int(num / 100)
            tens= int(num / 10) % 10
            units= num % 10

            if units % 2 != 0:
                continue

            freq[hundreds] -= 1
            if freq[hundreds] >= 0:
                freq[tens] -= 1

                if freq[tens] >= 0:
                    freq[units] -= 1

                    if freq[units] >= 0:
                        ans.append(num)
                    freq[units] += 1
                freq[tens] += 1
            freq[hundreds] += 1

        return ans
